implementation of motor thermistor support.

rowanG077 opened the pull request:
I have started the implementation of allowing users to connect a thermistor that is attached to a motor. The basic steps I took was create a new abstract class which allows current limiting. And based on that create a thermistor current limiter. The FET thermistor has been moved over to this class as well.

I have added some functions to get the adc_channel given a GPIO port/pin. User can choose which pins to use for the motor thermistor as long as it's a valid adc_channel. To differentiate the internal fet thermistor configuration contains a flag to indicate that the fet thermistor are on board and thus the pin assignment is immutable.

todo:

- [x] Create function in odrivetool to calculate and set thermistor coefficients based on thermistor constants.
- [x] Test in hardware

samuelsadok commented on Firmware/MotorControl/current_limiter.hpp:
```diff
@@ -0,0 +1,89 @@
+class ThermistorCurrentLimiter : public CurrentLimiter {
+public:
+    struct Config_t {
+        std::array<float, 4> thermistor_poly_coeffs = { };
```
Feels a bit scary to store an STL class to NVM. We don't have a guarantee that it doesn't contain any funny run-time variables. I'd prefer a classical C array.

rowanG077 replied:
According to cppreference it's guaranteed that `std::array` only non-static data member is semantically a plain C array.
https://en.cppreference.com/w/cpp/container/array

But your point stands. It's more obviously correct to just have a C-style array here.

samuelsadok replied:
Ah interesting, I didn't know that

samuelsadok commented on Firmware/MotorControl/current_limiter.hpp:
```diff
@@ -0,0 +1,89 @@
+                make_protocol_property("idx_1", &config_.thermistor_poly_coeffs[1]),
+                make_protocol_property("idx_2", &config_.thermistor_poly_coeffs[2]),
+                make_protocol_property("idx_3", &config_.thermistor_poly_coeffs[3])
+            ),
```
For `odrivetool` to properly back up the configuration, all NVM-stored config variables need to be exposed as `[...].config.[...]`.

Also as a heads-up: With #418, either `thermistor_poly_coeffs` would need to be turned into a class (like [`timing_log_`](https://github.com/madcowswe/ODrive/blob/feature/interface_autogen/Firmware/MotorControl/motor.hpp#L123-L127)) or the variables would have to be named `poly_coefficients_idxN` instead of `poly_coefficients.idxN` (although the latter naming is cleaner). Just saying this so we can avoid API breakage when #418 gets merged.

rowanG077 replied:
Will fix. I will rename the variables.

samuelsadok commented on Firmware/MotorControl/current_limiter.hpp:
```diff
@@ -0,0 +1,89 @@
+        return make_protocol_member_list(
+            make_protocol_property("error", &error_),
+            make_protocol_function("get_temp", *this, &ThermistorCurrentLimiter::get_temp),
+            make_protocol_function("set_pin", *this, &ThermistorCurrentLimiter::set_pin, "GPIO pin number to use for thermistor."),
```
This should probably be a setter hook function of the `config.gpio_pin` property, which should be read-write. Example: [`abs_spi_cs_gpio_pin`](https://github.com/madcowswe/ODrive/blob/devel/Firmware/MotorControl/encoder.hpp#L150-L151).
(btw the string is intended to act as argument name rather than an argument description)

rowanG077 replied:
Will fix.

samuelsadok commented on Firmware/MotorControl/motor.hpp:
```diff
@@ -185,7 +186,8 @@ class Motor {
             make_protocol_property("phase_current_rev_gain", &phase_current_rev_gain_),
             make_protocol_ro_property("thermal_current_lim", &thermal_current_lim_),
-            make_protocol_function("get_inverter_temp", *this, &Motor::get_inverter_temp),
+            make_protocol_object("fet_thermistor", fet_thermistor_.make_protocol_definitions()),
```
We should add a new section "API migration notes" to CHANGELOG.md to keep track of these changes.

rowanG077 replied:
Will do.

samuelsadok commented on tools/odrive/utils.py:
```diff
@@ -31,6 +32,25 @@
 class OperationAbortedException(Exception):
     pass
 
+def set_motor_thermistor_coeffs(motor, Rload, R_25, T_25, Beta, Tmin, TMax):
```
Maybe add a brief note somewhere in the `docs/` that describes how to set up the thermistor.

rowanG077 replied:
Will do.

samuelsadok commented on Firmware/MotorControl/low_level.cpp:
```diff
@@ -63,7 +63,7 @@ static uint16_t GPIO_port_samples [2][num_GPIO];
 *     Timer2.CCR3 (counter compare register 3)
 *     Timer2.CCR4 (counter compare register 4)
-* 
+*
```
There are a couple of whitespace-only changes. While they are generally sensible we should do this in a separate formatting PR (and also enforce it by CI) to avoid noise here.

rowanG077 replied:
Yes it was done automatically by my editor. I will pull these changes out.

samuelsadok commented on Firmware/MotorControl/current_limiter.cpp:
```diff
@@ -0,0 +1,84 @@
+    // We set the gpio_pin_ to UINT16_MAX so it's clear that this
+    // value is not a "real" gpio pin.
+    if (config_.onboard)
+    {
```
Our convention is to keep braces on the same line. See [here](https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions) and [here](https://google.github.io/styleguide/cppguide.html#Conditionals).

rowanG077 replied:
Will fix.

samuelsadok reviewed (commented):
Thanks for the PR, looks promising! I added a couple of inline comments and here are two more general comments:

1. There are some conceptual differences between onboard and off board thermistors. The former is an implementation detail and doesn't need to be exposed on the API. But if so, it should likely be immutable. As such I think it makes sense to split either `ThermistorCurrentLimiter` or `ThermistorCurrentLimiter::Config_t` into two classes, one that samples and exposes a temperature and one that calculates current limit based on that. The `motor_thermistor`'s config would be exposed on the API but not the `fet_thermistor`'s config.


2. It would be good to be able to query `ThermistorCurrentLimiter` multiple times per control cycle and get consistent results. This could be done by having a separate sampling function that stores the temperature in a member variable. So the interface I'd suggest would be:

       class ThermistorCurrentLimiter {
           void update(); // updates the `temp_` member variable based on ADC. Called from `Axis::do_updates()`. Not needed if latching is already done in a Thermistor class.
           bool do_checks(); // updates the `error_` member variable based on `temp_`. Called from `Axis::do_checks()`. Can also be assimilated into `update()`.
           float get_current_limit(float user_limit); // pure function, no member variables are changed.
       };

madcowswe commented on tools/odrive/utils.py:
```diff
@@ -31,6 +32,25 @@
+    motor.motor_thermistor.poly_coefficients.idx_0 = float(fit_temps[0])
+    motor.motor_thermistor.poly_coefficients.idx_1 = float(fit_temps[1])
+    motor.motor_thermistor.poly_coefficients.idx_2 = float(fit_temps[2])
+    motor.motor_thermistor.poly_coefficients.idx_3 = float(fit_temps[3])
```
I think the polynomial coefficients are in `fit` or `p1`, not in fit_temps.

madcowswe replied:
Either way, to avoid code duplication, we should have thermistors.py call this function (or other way around)

samuelsadok commented on Firmware/MotorControl/axis.cpp:
```diff
@@ -20,13 +30,23 @@ Axis::Axis(int axis_num,
       motor_(motor),
-      trap_(trap)
+      trap_(trap),
+      current_limiters_(make_array(
+          static_cast<CurrentLimiter*>(&fet_thermistor),
+          static_cast<CurrentLimiter*>(&motor_thermistor))),
```
as far as I can see you can do the same without `make_array`:
`current_limiters_{static_cast<CurrentLimiter*>(&fet_thermistor), static_cast<CurrentLimiter*>(&motor_thermistor)}`

rowanG077 replied:
@samuelsadok Yes you can but you lose some safety. For instance the following will compile without any warnings:

```
class Foo {
    Foo() : content({"test", "test2"}) {}
  
    std::array<std::string, 3> content;
};
```

The `make_array` helper ensures that you are forced to fill in every element else a compilation error is given. If you prefer it with a normal initializer_list I will change it.

samuelsadok replied:
Ah I see. I assume in this example the default constructor would be used for the third string. Then perhaps in our case it makes sense to check for `nullptr` in the loops that iterate over these arrays. Either that or `make_array` can be moved to "utils.hpp".

samuelsadok commented on Firmware/MotorControl/thermistor.hpp:
```diff
@@ -0,0 +1,132 @@
+    ThermistorCurrentLimiter(const ThermistorCurrentLimiter& other) = delete;
+    ThermistorCurrentLimiter(ThermistorCurrentLimiter&& other) noexcept = delete;
+    ThermistorCurrentLimiter& operator=(const ThermistorCurrentLimiter& original) = delete;
+    ThermistorCurrentLimiter& operator=(ThermistorCurrentLimiter&& original) = delete;
```
IMO for brevity we can omit these declarations. I mean it's true that it's not allowed to copy/move a `ThermistorCurrentLimiter` object but in other classes (Motor, Axis, ...) we also don't state this explicitly. Btw I think copy/move assignment operators are not autogenerated for classes that contain references (because the reference itself is const).

rowanG077 replied:
Fixed

samuelsadok commented on Firmware/MotorControl/thermistor.hpp:
```diff
@@ -0,0 +1,132 @@
+        uint16_t gpio_pin = 4;
+        float temp_limit_lower = 100;
+        float temp_limit_upper = 120;
+        bool enabled = false;
```
Technically a thermistor can just be disabled by putting the limits to infinity. On the other hand there is some merit to having an explicit enable flag. In any case we should make it consistent between onboard and offboard thermistors.

rowanG077 replied:
I didn't include an enable for onboard thermistors because I didn't think we should support disabling them. An enable flag is definitely needed for the offboard thermistors because else we possibly sample unconnected IO and create bogus temps that then trigger the temperature limits.

rowanG077 replied:
Fixed. Added enable to onboard thermistors as well.

samuelsadok commented on Firmware/MotorControl/thermistor.cpp:
```diff
@@ -0,0 +1,86 @@
+}
+
+bool ThermistorCurrentLimiter::do_checks() {
+    if (enabled_ && temp_ >= temp_limit_upper_) {
```
We should keep the 5°C margin here that was in the old code to account for measurement noise.

rowanG077 replied:
Fixed

samuelsadok commented on Firmware/MotorControl/thermistor.cpp:
```diff
@@ -0,0 +1,86 @@
+{
+}
+
+float ThermistorCurrentLimiter::get_temp() const {
```
I think this is not used anymore

rowanG077 replied:
Fixed.

samuelsadok reviewed (commented):
Looks good, just a few minor comments.

If you want to build the HTML docs you can check out [this](https://docs.odriverobotics.com/developer-guide#documentation). But for most purposes a simple markdown previewer will do (I'm using the VSCode extension "Markdown All in One" by Yu Zhang).

Yes PRs should go to `devel`, sorry I didn't notice.

samuelsadok commented on docs/thermistors.md:
```diff
@@ -0,0 +1,35 @@
+# Thermistors
```
You can add this file to "docs/_data/index.yaml" so it will show up in the navigation bar.

rowanG077 replied:
Fixed

samuelsadok commented on Firmware/odrive-interface.yaml:
```diff
@@ -211,6 +211,8 @@ interfaces:
           MinEndstopPressed:
           MaxEndstopPressed:
+          OverTemp:
+            doc: Check `fet_thermistor.error` and `motor_thermistor.error` for more information.
```
let's move this after `HomingWithoutEndstop` to reduce the likelihood of breaking someone's code silently

rowanG077 replied:
Fixed.

samuelsadok commented on Firmware/odrive-interface.yaml:
```diff
@@ -337,7 +380,6 @@ interfaces:
           UnexpectedTimerCallback:
           CurrentSenseSaturation:
-          InverterOverTemp:
           CurrentLimitViolation:
```
again to reduce the likelihood of silent errors you can "pin" the bit number of the succeeding flag: `CurrentLimitViolation: {bit: 12}`

rowanG077 replied:
Fixed.

samuelsadok commented on Firmware/odrive-interface.yaml:
```diff
@@ -656,6 +694,12 @@ valuetypes:
         doc: run axis homing function
 
+  ODrive.ThermistorCurrentLimiter.Error:
+    values:
```
For this enum the `flags` key should be used instead of `values` to make the interface generator chose numbers that are powers of two (see for example `ODrive.Motor.Error`). The `nullflag` key can then be used to specify the name of the "no flag is set"-state.

rowanG077 replied:
Fixed.

samuelsadok reviewed (commented):
Cool thanks!

Yes current limits make more sense when talking about thermal clamping. The `current_lim` is still exposed to the user and internally `effective_current_lim()` is still being honored so as far as I can see the new torque limit doesn't require any changes to this PR.

I added some minor comments on the YAML changes but apart from that it looks good to me. Regarding testing I think you covered all important aspects, so no further requests from my side.
I think @madcowswe also wants to take a look so give him a few days.

madcowswe commented on docs/thermistors.md:
```diff
@@ -0,0 +1,35 @@
+Thermistors are elements that change their resistance based on the temperature. They can be used to electrically measure temperature. The ODrive itself has thermistors on board near the FETs to ensure that they don't burn themselves out. In addition to this it's possible to connect your own thermistor to measure the temperature of the connected motors. There are two types of thermistors, Negative ...
+
+## FET thermistor
+The temperature of the onboard FET thermistors can be read out by using the `odrivetool` under `<axis>.fet_thermistor.temp`. The odrive will automatically start current limiting the motor when the `<axis>.fet_thermistor.config.temp_limit_lower` threshold is exceeded and once `<axis>.fet_thermistor.config.temp_limit_upper` is exceeded the ODrive will stop controlling the motor and set an error. Th ...
```
small grammar edit: "can be changed values but this is not recommended" -> "can be changed, but this is not recommended"

rowanG077 replied:
Fixed.

madcowswe commented on docs/thermistors.md:
```diff
@@ -0,0 +1,35 @@
+
+* `gpio_pin`: The GPIO input in used for this thermistor.
+* `poly_coefficient_0` to `poly_coefficient_3`: Coefficient that needs to be set for your specific setup more on that in [Thermistor coefficients](#Thermistor coefficients).
+* `temp_limit_lower` and `temp_limit_upper`: Same as the FET temperature limits.
```
"same principle" or "same idea" (so they don't think they need to put the same values)

rowanG077 replied:
Fixed.

madcowswe commented on docs/thermistors.md:
```diff
@@ -0,0 +1,35 @@
+
+The way this works is that the thermistor is connected in series with a known resistance value. By connecting an ADC directly after the thermistor the resistance value can be determined. For further information see [Voltage divider](https://en.wikipedia.org/wiki/Voltage_divider).
+
+To use a thermistor with the ODrive a voltage divider circuit has to be made that uses `VCCA` as the power source with `GNDA` as the ground. The voltage divider output can be connected to a GPIO pin that supports analog input.
```
add: "While not strictly necessary, it is a good idea to add a capacitor as shown as well. This will help reduce the effect of electrical noise. A value between 470nF and 4.7uF is recommended, and any voltage rating 4V or higher. Put the capacitor physically close to the ODrive."

rowanG077 replied:
Fixed.

madcowswe commented on docs/thermistors.md:
```diff
@@ -0,0 +1,35 @@
+* `Rload`: The Ohm value of the resistor used in the voltage divider circuit. 
+* `R_25`: The resistance of the thermistor when the temperature is 25 degrees celsius. Can usually be found in the datasheet of your thermistor. Can also be measured manually with a multimeter.
+* `Beta`: A constant specific to your thermistor. Can be found in the datasheet of your thermistor.
+* `Tmin` and `Tmax`: The minimum and maximum temperature limit that is used to create the coefficients. Make sure to set these values that are lower and higher respectively then what is expected during operation.
```
"The temperature range that is used to create the coefficients. Make sure to set this range to be wider than what is expected during operation. A good example may be -10 to 150."

rowanG077 replied:
Fixed.

madcowswe commented on Firmware/MotorControl/motor.cpp:
```diff
@@ -186,10 +162,15 @@ float Motor::effective_current_lim() {
-    return current_lim;
+    // Apply axis current limiters
+    for (const CurrentLimiter* const limiter : axis_->current_limiters_) {
+        current_lim = std::min(current_lim, limiter->get_current_limit(config_.current_lim));
```
I found it a bit confusing that get_current_limit takes an argument. I think the purpose being a "default" if it isn't enabled? If so then this min is redundant right?
Supposing I understand correctly, then I think it might be clearer for get_current_limit to not take any argument and return an optional https://en.cppreference.com/w/cpp/utility/optional

rowanG077 replied:
The current limiter doesn't "know" the configured current limit. The purpose of the argument is not only being a default but also what base current limit to apply the limit to. An alternative would be for current limiters to not return a concrete current value but rather a current scale between 0.0 and 1.0. And then the logic would simply be to find the smallest scaler and apply it to the configured current limit.

An optional would be overkill I think because we would need the `min` call anyway. Suppose we have two current limiters with a configured limit of 20A.

1. First current limiter says current limit is 15A.
2. Second current limiter says current limit is 12A.

Then we would still need to find the minimum of all the current limits to know which to actually apply. Introducing an optional is simply another case without a (to me) clear benefit. 

Anyway the best way forward I think would be to rewrite this to a scaler based implementaiton. `get_current_limit` would not have any arguments and return a float between 0.0 and 1.0. The `std::min` would remain and the `effective_current_lim_` will become `current_lim * current_lim_scaler`.

madcowswe replied:
Ah ok, the value is also used as a base current limit to scale from, I see.
Now that I understand what is going on, I think it is good as is. Changing it to using `current_lim_scaler` will preclude other limiting scenarios that are an absolute limit rather than ratiometric de-rating.

Perhaps we can rename it to `base_current_lim` instead of user limit, it may make it a bit clearer that the main purpose is to do scaling off, rather than a user default (which is what I misinterpreted it as).

rowanG077 replied:
Fixed.

madcowswe commented on Firmware/MotorControl/thermistor.hpp:
```diff
@@ -0,0 +1,74 @@
+    uint16_t adc_channel_;
+    const float* const coefficients_;
+    const size_t num_coeffs_;
+    float temp_;
```
just "temp" by itself can look like "temporary". I would suggest using "temperature" in this case. Also applies to the interface yaml.

rowanG077 replied:
Fixed.

madcowswe reviewed (commented):
Great work, looks very nice in general.
I have some small comments on some wording, and one suggested code change.

Thanks for your work on this!

rowanG077 committed:
- Refactored thermal and current lmiting sub-systems and introduced motor thermistor support.
- Added thermal errors to dump_errors call in odrivetool.
- Added function in odrivetool to set thermistor coefficients based on thermistor specs.
- Added documentation for users on how to connect and configure their own thermistors.
